subsys/bluetooth/audio/has.c

```diff
@@ -34,7 +56,362 @@ static ssize_t read_features(struct bt_conn *conn, const struct bt_gatt_attr *at
+		client->params.ind.destroy = NULL;
+		client->params.ind.data = buf->data;
+		client->params.ind.len = buf->len;
+
+		return bt_gatt_indicate(client->conn, &client->params.ind);
+	}
+
+	return -ECANCELED;
+}
+
+static int bt_has_cp_read_preset_rsp(struct has_client *client, struct bt_has_preset *preset,
+				     uint8_t is_last)
+{
+	struct bt_has_cp_hdr *hdr;
+	struct bt_has_cp_read_preset_rsp *rsp;
+
+	NET_BUF_SIMPLE_DEFINE(buf, sizeof(*hdr) + sizeof(*rsp) + BT_HAS_PRESET_NAME_MAX);
+
+	BT_DBG("conn %p preset %p is_last 0x%02x", (void *)client->conn, preset, is_last);
+
+	hdr = net_buf_simple_add(&buf, sizeof(*hdr));
+	hdr->opcode = BT_HAS_OP_READ_PRESET_RSP;
+	rsp = net_buf_simple_add(&buf, sizeof(*rsp));
+	rsp->is_last = is_last;
+	rsp->index = preset->index;
+	rsp->properties = preset->properties;
+	net_buf_simple_add_mem(&buf, preset->name, strlen(preset->name));
+
+	return control_point_send(client, &buf);
+}
+
+static uint8_t preset_find_next(const struct bt_has_preset *preset, void *user_data)
+{
+	const struct bt_has_preset **found = user_data;
+
+	if (*found == NULL || (*found)->index > preset->index) {
+		*found = preset;
+	}
+
+	return BT_HAS_PRESET_ITER_CONTINUE;
+}
+
+static void process_control_point_work(struct k_work *work)
+{
+	struct has_client *client = CONTAINER_OF(work, struct has_client, control_point_work);
+	int err;
+
+	if (!client->conn) {
+		return;
+	}
+
+	if (read_presets_req_is_pending(client)) {
+		struct bt_has_preset *preset = NULL;
+		const uint8_t start_index = client->read_presets_req.start_index;
+		uint8_t is_last;
+
+		preset_foreach(start_index, BT_HAS_PRESET_INDEX_LAST, preset_find_next, &preset);
+
+		if (unlikely(preset == NULL)) {
+			(void)bt_has_cp_read_preset_rsp(client, NULL, 0x01);
+
+			return;
+		}
+
+		if (client->read_presets_req.num_presets > 1) {
+			struct bt_has_preset *next = NULL;
+
+			preset_foreach(preset->index + 1, BT_HAS_PRESET_INDEX_LAST,
+				       preset_find_next, &next);
+
+			is_last = next == NULL ? 0x01 : 0x00;
+
+		} else {
+			is_last = 0x01;
+		}
+
+		err = bt_has_cp_read_preset_rsp(client, preset, is_last);
+		if (err) {
+			BT_ERR("bt_has_cp_read_preset_rsp failed (err %d)", err);
+		}
+
+		if (err || is_last) {
+			read_presets_req_free(client);
+		} else {
+			client->read_presets_req.start_index = preset->index + 1;
+			client->read_presets_req.num_presets--;
+		}
+	}
+}
+
+static uint8_t handle_read_preset_req(struct bt_conn *conn, struct net_buf_simple *buf)
+{
+	const struct bt_has_cp_read_presets_req *req;
+	struct bt_has_preset *preset = NULL;
+	struct has_client *client;
+
+	if (buf->len < sizeof(*req)) {
+		return BT_HAS_ERR_INVALID_PARAM_LEN;
+	}
+
+	/* As per HAS_d1.0r00 Client Characteristic Configuration Descriptor Improperly Configured
+	 * shall be returned if client writes Read Presets Request but is not registered for
+	 * indications.
+	 */
+	if (!bt_gatt_is_subscribed(conn, &has_svc.attrs[4], BT_GATT_CCC_INDICATE)) {
```

Comment:
Given that `&has_svc.attrs[4]` is access multiple places, perhaps a macro or some other named way of getting this specific attribute would be useful